### PR TITLE
[CI] Enable pre-commit testing on Windows

### PR DESCRIPTION
**Pull request by uditagarwal97**
This PR (1) enables running Windows Build + LIT + E2E tests during pre-commit testing, (2) Fix windows_test_comment_trigger workflow after #8656

### CODE REVIEW

File: .github/workflows/sycl_precommit.yml [hunk 1]
```diff
@@ -66,3 +66,12 @@ jobs:
...
       lts_aws_matrix: ${{ needs.test_matrix.outputs.lts_aws_matrix }}
+
+  windows_default:
+    name: Windows
+    needs: test_matrix
+    if: github.repository == 'intel/llvm'
+    uses: ./.github/workflows/sycl_windows_build_and_test.yml
+    with:
+      lts_matrix: ${{ needs.test_matrix.outputs.lts_wn_matrix }}
+      build_ref: ${{ github.event.pull_request.head.sha }}
```
Comment (aelovikov-intel) [inline .github/workflows/sycl_precommit.yml:77]: I see that it follows Linux path, but maybe `github.ref` could be a better choice:

> The fully-formed ref of the branch or tag that triggered the workflow run. For workflows triggered by push, this is the branch or tag ref that was pushed. For workflows triggered by pull_request, this is the pull request merge branch. For workflows triggered by release, this is the release tag created. For other triggers, this is the branch or tag ref that triggered the workflow run. This is only set if a branch or tag is available for the event type. The ref given is fully-formed, meaning that for branches the format is refs/heads/<branch_name>, for pull requests it is refs/pull/<pr_number>/merge, and for tags it is refs/tags/<tag_name>. For example, refs/heads/feature-branch-1.

Anyway, could/should be a separate PR if at all.

File: .github/workflows/windows_test_comment_trigger.yml [hunk 2]
```diff
@@ -42,14 +42,22 @@ jobs:
     outputs:
       PR_SHA: ${{ steps.sha.outputs.result }}
 
+  test_matrix:
+    name: Generate Test Matrix
+    if: github.repository == 'intel/llvm'
+    uses: ./.github/workflows/sycl_gen_test_matrix.yml
+    with:
+      lts_config: "l0_gen9;win_l0_gen12"
```
Comment (aelovikov-intel) [inline .github/workflows/windows_test_comment_trigger.yml:50]: Why do we have Linux task ("l0_gen9") here?

Comment (uditagarwal97) [reply .github/workflows/windows_test_comment_trigger.yml]: Perhaps, we don't need it. I've fixed it.

File: .github/workflows/sycl_precommit.yml (hunk 1 above)
Comment (uditagarwal97) [reply .github/workflows/sycl_precommit.yml:77]: Agree. Perhaps a separate PR for fixes to both Linux and Windows jobs would be better.

**Review by bader — CHANGES_REQUESTED** [review]
Having both configurations: Windows pre-commit and Windows testing on demand doesn't make sense to me.
We added on-demand Windows testing because according to our estimates CI doesn't have enough resources to handle the load added by Windows pre-commit. Both jobs do the same thing, but pre-commit launches automatically for each commit. On-demand testing is not needed if each commit is tested.

I have no objections to run an experiment with enabling Windows in pre-commit to check how CI will handle the load.

There should be only on configuration, so if you go with pre-commit, please, remove or disable "on-demand" testing.

@romanovvlad, FYI.

**Comment by uditagarwal97** [comment]
@bader Agree. I've temporarily disabled the "on-demand" windows testing workflow. We can remove it completely, let's say after a week, if the Windows CI handles the load as we expect.

File: .github/workflows/windows_test_comment_trigger.yml [hunk 3]
```diff
@@ -7,7 +7,7 @@ on:
 jobs:
   windows_test_preparation:
     runs-on: ubuntu-20.04
-    if: ${{ github.event.issue.pull_request && github.event.comment.body == '/testwin' }}
+    if: ${{ false && github.event.issue.pull_request && github.event.comment.body == '/testwin' }}
```
Comment (bader) [inline .github/workflows/windows_test_comment_trigger.yml:10]: Disabling part of the jobs doesn't seem like a good solution. I would rather disable the whole workflow either manually (after this PR merge) or programmatically modifying the lines 3-5.
I also talked to @romanovvlad and we agreed that windows_test_completion should also be skipped if `windows_test_preparation` and `windows_default` jobs are skipped.

Most of the time the workflow execution looks like this - https://github.com/intel/llvm/actions/runs/4633598850. We run only `windows_test_completion`, which doesn't do anything useful as most of the comments are about reviewing the code rather than running the tests on windows.

File: .github/workflows/windows_test_comment_trigger.yml [hunk 4]
```diff
@@ -42,14 +42,22 @@ jobs:
     outputs:
       PR_SHA: ${{ steps.sha.outputs.result }}
 
+  test_matrix:
+    name: Generate Test Matrix
+    if: ${{ false && github.repository == 'intel/llvm' }}
+    uses: ./.github/workflows/sycl_gen_test_matrix.yml
+    with:
+      lts_config: "win_l0_gen12"
+      cts_config: "cuda"
```
Comment (bader) [inline .github/workflows/windows_test_comment_trigger.yml:51]: Please, remove this config. We don't have Windows machines with NVIDIA GPU.

File: .github/workflows/windows_test_comment_trigger.yml [hunk 5]
```diff
@@ -42,14 +42,22 @@ jobs:
     outputs:
       PR_SHA: ${{ steps.sha.outputs.result }}
 
+  test_matrix:
+    name: Generate Test Matrix
```
Comment (bader) [inline .github/workflows/windows_test_comment_trigger.yml:46]: We don't need to run this job for all comments - only for those which trigger windows tests.
Please, add `needs: [windows_test_preparation]`.

File: .github/workflows/windows_test_comment_trigger.yml (hunk 4 above)
Comment (uditagarwal97) [reply .github/workflows/windows_test_comment_trigger.yml]: Done.

File: .github/workflows/windows_test_comment_trigger.yml (hunk 3 above)
Comment (uditagarwal97) [reply .github/workflows/windows_test_comment_trigger.yml]: Okay, you can manually disable the workflow after the PR merge.
Also, regarding `windows_test_completion` workflow, I've removed the `if: always()` statement so this workflow will also gets skipped if the `windows_default` and `windows_test_preparation` workflows are skipped.

Comment (bader) [reply .github/workflows/windows_test_comment_trigger.yml]: > Also, regarding `windows_test_completion` workflow, I've removed the `if: always()` statement so this workflow will also gets skipped if the `windows_default` and `windows_test_preparation` workflows are skipped.

FYI. It was added for reporting the status of failed jobs (see https://github.com/intel/llvm/commit/a1fbd64649153161e40c00aa84cf923f14506266). I think the right way to fix this is to replace `needs: [windows_test_preparation, windows_default]` with something like `if: jobs.windows_test_preparation.result != skipped` (syntax to be checked). Otherwise `needs:` will run the status reporting job only if testing was successful.

This is not a blocking issues as we are going to disable this job anyway, but if our experiments with pre-commit will be unsuccessful, we will need to fall back to on-demand testing and we need accurate reporting.

**Review by bader — APPROVED** [review]
Thanks for working on this!

**Review by aelovikov-intel — APPROVED** [review]

**Comment by uditagarwal97** [comment]
@intel/llvm-gatekeepers The PR is ready!

**Comment by AaronBallman** [comment]
Don't let me hold anything up in this space, but one thing to consider for Windows-specific precommit testing is to build a debug configuration so that we get checked & debug iterator testing (https://learn.microsoft.com/en-us/cpp/standard-library/safe-libraries-cpp-standard-library?view=msvc-170). That helps catch quite a few mistakes (I just fixed one this morning that turned out to be `A.find(X) != B.end()` where `A` and `B` are the same type but different objects -- MSVC debug iterators caught the issue immediately).

File: .github/workflows/windows_test_comment_trigger.yml (hunk 3 above)
Comment (bader) [reply .github/workflows/windows_test_comment_trigger.yml]: Sounds good to me. Thanks for clarifying.
